In [1]:
import pandas as pd
import numpy as np

In [3]:
trainData = pd.read_csv("../data/train.csv")
testData = pd.read_csv("../data/test.csv")
validationData = pd.read_csv("../data/validation.csv")

print(f"Shape of trainData is : {trainData.shape}")
print(f"Shape of testData is : {testData.shape}")
print(f"Shape of validationData is : {validationData.shape}")
trainData.head()

Shape of trainData is : (7827, 14)
Shape of testData is : (967, 14)
Shape of validationData is : (870, 14)


,Unnamed,team,opp,host,year,month,toss,day_match,bat_first,format,fow,score,rpo,result
0,4901,australia,south_africa,sri_lanka,2012,sep,1,0,0,1,5,146,7.30,1
1,1910,india,australia,india,2020,jan,0,0,1,0,6,340,6.80,1
2,5735,canada,scotland,scotland,2009,jul,1,1,0,0,4,286,5.72,1
3,2136,australia,england,australia,1987,jan,1,1,1,0,6,225,4.50,1
4,4371,new_zealand,pakistan,uae,2009,nov,0,0,0,1,5,153,7.65,0


In [4]:
import sys
import os
sys.path.insert(0, os.path.abspath('..'))  # Adjust path as needed

from utils import pre_process

# Step 2: Encode the columns["team", "opp", "host", "month"]
categorical_cols = ["team", "opp", "host", "month"]
continuous_cols = ["Unnamed", "year", "fow", "score", "rpo"]

trainDataX, trainDataY, validationDataX, validationDataY, testDataX, testDataY =  pre_process(trainData, validationData, testData, categorical_cols = categorical_cols, continuous_cols = continuous_cols)
trainDataX.head()

After pre-processing
Shape of trainDataX is : (7827, 13)
Shape of trainDataY is : (7827,)
Shape of testDataX is : (967, 13)
Shape of testDataY is : (967,)
Shape of validationDataX is : (870, 13)
Shape of validationDataY is : (870,)


,Unnamed,team,opp,host,year,month,toss,day_match,bat_first,format,fow,score,rpo
0,0.023176,0,0,0,0.527774,0,1,0,0,1,-1.226069,-1.161778,0.976509
1,-1.050706,1,1,1,1.215295,1,0,0,1,0,-0.742084,1.817869,0.696053
2,0.322613,2,2,2,0.269953,2,1,1,0,0,-1.710053,0.988483,0.090266
3,-0.969563,0,3,3,-1.620731,1,1,1,1,0,-0.742084,0.051583,-0.594049
4,-0.167114,3,4,4,0.269953,3,0,0,0,1,-1.226069,-1.054265,1.172829


In [5]:
(trainDataX == -1).any().any(), (testDataX == -1).any().any(), (validationDataX == -1).any().any()

(False, False, False)

In [6]:
trainDataX[trainDataX["team"] == 0].head()

,Unnamed,team,opp,host,year,month,toss,day_match,bat_first,format,fow,score,rpo
0,0.023176,0,0,0,0.527774,0,1,0,0,1,-1.226069,-1.161778,0.976509
3,-0.969563,0,3,3,-1.620731,1,1,1,1,0,-0.742084,0.051583,-0.594049
10,-0.874059,0,3,1,-1.448850,7,1,1,1,0,-2.194037,0.312686,-0.403338
15,-0.794712,0,0,8,0.269953,9,0,0,1,0,-0.258100,1.249586,0.280976
18,0.280965,0,7,5,0.269953,6,1,1,1,1,-0.258100,-0.808521,1.621560


In [7]:
# t20 -> format = 1
# odi -> format = 0
print("Unique counts")
for column in trainData.columns:
    print(f"{column} :: {len(trainData[column].unique())}")

Unique counts
Unnamed :: 7827
team :: 19
opp :: 19
host :: 19
year :: 53
month :: 12
toss :: 2
day_match :: 3
bat_first :: 2
format :: 2
fow :: 10
score :: 349
rpo :: 622
result :: 2


In [8]:
# just playing around with data
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

team_summary = trainData.groupby(["team","opp", "host", "toss","result"]).size().unstack(fill_value=0)
team_summary.head(2000)

result                                        0   1
team         opp          host         toss        
afganistan   bangladesh   australia    0      1   0
                          bangladesh   0      8   3
                                       1      1   3
                          india        0      0   2
                                       1      0   1
                          pakistan     0      1   0
                          uae          0      1   1
                                       1      0   1
             canada       canada       1      0   2
                          netherlands  0      0   1
                          sri_lanka    0      0   1
                          uae          0      1   1
                                       1      0   1
             ireland      india        0      0   8
                                       1      2   3
                          ireland      0      1   6
                                       1      5   2
                          netherlands  0      1   0
                          uae          0      2   7
                                       1      1   1
                          zimbambave   0      0   1
             kenya        kenya        0      1   1
                          netherlands  0      0   1
                          uae          0      1   1
                                       1      0   2
             netherlands  netherlands  0      0   2
                                       1      1   0
                          scotland     0      0   1
                          uae          0      0   1
                                       1      2   1
             scotland     india        1      0   1
                          netherlands  1      1   0
                          new_zealand  1      0   1
                          scotland     0      0   1
                                       1      1   2
                          south_africa 1      0   1
                          uae          0      1   1
                                       1      0   3
                          zimbambave   0      1   0
             sri_lanka    australia    1      1   0
                          pakistan     0      1   0
                          sri_lanka    0      1   0
                                       1      1   1
             uae          bangladesh   0      1   0
                          malaysia     0      0   1
                          scotland     0      0   1
                          uae          0      1   1
                                       1      2   3
             zimbambave   bangladesh   0      0   1
                                       1      1   0
                          india        1      0   1
                          uae          0      2   6
                                       1      0   6
                          zimbambave   0      3   6
                                       1      3   7
australia    canada       india        0      0   1
             england      australia    0      9  19
                                       1      8  23
                          england      0     19  19
                                       1     15  11
                          india        1      1   2
                          south_africa 0      0   3
                          uae          1      1   1
                          west_indis   0      1   1
             india        australia    0      6  12
                                       1      7  18
                          bangladesh   1      1   0
                          england      0      2   1
                                       1      0   1
                          india        0     13  10
                                       1     14  21
                          malaysia     1      0   1
                          new_zealand  0      0   1
                                       1      0   1
                          south_africa 0      0 

In [9]:
# Question A

# now build a decision tree to predict the result given the trainingData
# Step 0: 
from decisionTree import DecisionTreeModel
from utils import evaluate, plotA

maxDepths = [5, 10, 15, 20]
trainAccuracies = []
validationAccuracies = []
testAccuracies = []

categorical_cols = ["team", "opp", "host", "month", "toss", "day_match", "bat_first", "format"]
continuous_cols = ["Unnamed", "year", "fow", "score", "rpo"]

for maxDepth in maxDepths:
    print(f"---------------Testing for maxDepth : {maxDepth}-------------------")
    model = DecisionTreeModel(maxDepth, categorical_cols = categorical_cols, continuous_cols = continuous_cols)

    # TRAIN THE MODEL WITH TRAIN DATA
    print("Training on the train data")
    model.fit(trainDataX, trainDataY)

    # if maxDepth == 5:
    #     model.print_tree()
    
    # PREDICT THE MODEL ON TRAIN DATA
    print("Prediction on train data")
    predictY_train = model.predict(trainDataX)
    
    # EVALUATION
    train_accuracy = evaluate(predictY_train, trainDataY)
    
    # PREDICT THE MODEL ON VALIDATION DATA
    print("Prediction on validation data")
    predictY_validation = model.predict(validationDataX)
    
    # EVALUATION
    validation_accuracy = evaluate(predictY_validation, validationDataY)
    
    # PREDICT THE MODEL ON Test DATA
    print("Prediction on test data")
    predictY_test = model.predict(testDataX)
    
    # EVALUATION
    test_accuracy = evaluate(predictY_test, testDataY)

    trainAccuracies.append(train_accuracy)
    validationAccuracies.append(validation_accuracy)
    testAccuracies.append(test_accuracy)

plotA(maxDepths, trainAccuracies, validationAccuracies, testAccuracies)

---------------Testing for maxDepth : 5-------------------
Training on the train data
Prediction on train data
Accuracy : 0.889, Precision: 0.910, Recall: 0.866, F1-Score: 0.888
Prediction on validation data
Accuracy : 0.583, Precision: 0.587, Recall: 0.553, F1-Score: 0.569
Prediction on test data
Accuracy : 0.582, Precision: 0.557, Recall: 0.572, F1-Score: 0.565
---------------Testing for maxDepth : 10-------------------
Training on the train data
Prediction on train data
Accuracy : 0.995, Precision: 0.999, Recall: 0.991, F1-Score: 0.995
Prediction on validation data
Accuracy : 0.606, Precision: 0.605, Recall: 0.606, F1-Score: 0.605
Prediction on test data
Accuracy : 0.589, Precision: 0.563, Recall: 0.598, F1-Score: 0.580
---------------Testing for maxDepth : 15-------------------
Training on the train data
Prediction on train data
Accuracy : 0.996, Precision: 0.999, Recall: 0.992, F1-Score: 0.996
Prediction on validation data
Accuracy : 0.606, Precision: 0.605, Recall: 0.606, F1-Scor

TypeError: plotA() missing 1 required positional argument: 'file_path'

| Max Depth | Train Accuracy | Validation Accuracy | Test Accuracy |
| :-------- | :------------- | :------------------ | :------------ |
| 5         | 0.889          | 0.585               | 0.583         |
| 10        | 0.995          | 0.608               | 0.592         |
| 15        | 0.996          | 0.608               | 0.592         |
| 20        | 0.996          | 0.608               | 0.592         |

When depth = 5,
The model underfits slightly — not enough complexity to capture all training patterns.
Training accuracy (0.889) < 1.0 indicates this.

As depth increases to 10 and beyond,

The tree almost perfectly memorizes training data (train accuracy → 0.995+).
However, test accuracy barely improves (~0.59) — showing overfitting.

After depth = 10, there’s no meaningful performance gain on test/validation.

General observation:
Increasing tree depth makes the model more expressive but reduces generalization.

Optimal depth seems to be around 10 for this dataset.

As the maximum depth of the decision tree increases, the model achieves nearly perfect accuracy on the training set, indicating overfitting. However, the validation and test accuracies remain almost constant around 0.592, suggesting that the model fails to generalize beyond a certain complexity level. The optimal depth appears to be around 10, beyond which deeper trees add no value to predictive performance.